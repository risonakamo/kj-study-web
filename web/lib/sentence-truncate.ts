// funcs for truncating sentence

import _ from "lodash";

/** split sentences on these symbols */
const SplitReg:RegExp=/[、。]/;

/** when truncating, keep above this size */
const SentenceMinSize:number=10;

const DefaultTruncatePercent:number=.5;

/** special jisho truncate */
export function jishoTruncate(input:string):string
{
    // split sentence into fragments based on selected set of symbols
    const fragments:string[]=input.split(SplitReg);

    // find the largest fragment
    const largestFragment:string|undefined=_.maxBy(fragments,
    (fragment:string):number=>{
        return fragment.length;
    });

    if (!largestFragment)
    {
        return "";
    }

    // try truncating the fragment to percentage size
    const truncated:string=percentTruncate(largestFragment,DefaultTruncatePercent);

    // take this truncated if it is above the min size
    if (truncated.length>SentenceMinSize)
    {
        return truncated;
    }

    // otherwise, return the largest fragment truncated to the min size (if it is
    // even that long)
    return largestFragment.slice(0,SentenceMinSize);
}

/** truncate a string to a percent length, rounded up. percent should be 0-1 */
function percentTruncate(input:string,size:number):string
{
    const newSize:number=Math.ceil(input.length*size);

    return input.slice(0,newSize);
}