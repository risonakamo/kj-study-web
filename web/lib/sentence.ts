// functions for manipulating sentences

/** given a sentence, split it into pieces on a target word. the word will remain in the split. */
export function splitSentenceOnWord(sentence:string,word:string):string[]
{
    return sentence.split(new RegExp(`(${word})`));
}