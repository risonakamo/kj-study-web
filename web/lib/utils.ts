import _ from "lodash";

/** add to end of a url a seed query. for iframe src change to work */
export function addUrlSeed(url:string):string
{
    const params=new URL(url);
    params.searchParams.append("rand",_.random(1,10000).toString());

    return params.toString();
}