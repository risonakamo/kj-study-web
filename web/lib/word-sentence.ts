// lib for manipulating word sentence data type
// deprecated for now, as backend sends in word sentence pair form already

// import _ from "lodash";

// /** convert word sentence dict to list of single word sentences */
// export function wordSentenceDictToSingles(sentenceDict:WordSentenceDict):SingleWordSentence[]
// {
//     return _.flatMap(sentenceDict,(sentences:string[],word:string):SingleWordSentence[]=>{
//         return _.map(sentences,(sentence:string):SingleWordSentence=>{
//             return {
//                 word,
//                 sentence
//             };
//         });
//     });
// }