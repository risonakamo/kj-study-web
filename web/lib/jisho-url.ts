// funcs dealing with jisho url scheme

import {percentTruncate} from "./utils";

/** percent to truncate sentence when searching. adjust if failing alot */
const SentenceTruncatePercent:number=.5;

/** return url to search for a target word */
export function searchForWordUrl(word:string):string
{
    return `https://jisho.org/search/${word}`;
}

/** return url to search for sentence. automatically truncates the sentence to
 *  some percent of the size for consistency */
export function searchForSentenceUrl(sentence:string):string
{
    sentence=percentTruncate(sentence,SentenceTruncatePercent);
    return `https://jisho.org/search/${sentence}%20%23sentence`;
}