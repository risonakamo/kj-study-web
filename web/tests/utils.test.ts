import {expect, test} from "vitest";

import {percentTruncate} from "@/lib/utils";
import _, {result} from "lodash";

test("percent-truncate",()=>{
    const inputs:[string,number][]=[
        ["something",.75],
        ["something",101],
        ["huh",.75],
        ["",.5],
        ["a",1],
        ["a",.5],
        ["a",.0000001],
        [" adlsn asdfnkas dfaksgd",0],
        [" adlsn asdfnkas dfaksgd",.25]
    ];

    const results:string[]=_.map(inputs,(input:[string,number]):string=>{
        const result:string=percentTruncate(input[0],input[1]);

        console.log(
            input,
            "->",
            result
        );

        return result;
    });

    expect(results).toMatchSnapshot();
});