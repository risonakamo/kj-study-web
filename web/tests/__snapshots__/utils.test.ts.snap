// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`percent-truncate 1`] = `
[
  "somethi",
  "something",
  "huh",
  "",
  "a",
  "a",
  "a",
  "",
  " adlsn",
]
`;
